server/src/main/java/org/elasticsearch/action/DocWriteRequest.java

```diff
@@ -256,4 +257,23 @@ static ActionRequestValidationException validateSeqNoBasedCASParams(
 
         return validationException;
     }
+
+    // A heuristic for the bytes overhead of a single indexing operation
+    int WRITE_REQUEST_BYTES_OVERHEAD = 256;
```

Comment:
I wonder if this hinders describing the limit in a sensible way in documentation. With a 4GB node and the defaults, this allows 3355443 simultaneous operations to be active (when source is empty). This makes me wonder if we should rather also introduce a limit on the number of operations and express the byte-size limit as only source. I think this will be easier to document and understand for our users. The default for such a count could depend on heap size or the limit setting.

We could also just up this factor to something like 1024.

This does tie in with removing the limit on the write thread-pool queue. As such, this PR adds an additional limit compared to current and that is fine, but I think the intention is to remove that limit and once that is done, I think the byte-sized limit is all we have before the real-memory circuit breaker kicks in?